### PR TITLE
post: P384 OpenSSL Explainer Blogpost

### PR DESCRIPTION
**Pull request by rohanmclure**
This is my blog post explaining the overhaul of the implementation used for `secp348r1`. If the maths makes any sense - even if the stages just make sense to follow (even if the algorithm and its correctness is a little hazy) - let me know, or otherwise how I can make it better. Thanks

### CODE REVIEW

**Comment by rohanmclure** [comment]
@ajdlinux @ruscur @BenjaminGrayNp1 @NMiehlbradt

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 1]
```diff
@@ -0,0 +1,214 @@
+Title: Going out on a Limb: Efficient Elliptic Curve Arithmetic in OpenSSL
+Authors: Rohan McLure
+Date: 2023-07-31 13:15
+Category: Cryptography
+
+So I've just submitted a pull request to OpenSSL for a new strategy I've been developing for efficient field arithmetic used in secp384r1, a curve prescribed by NIST for digital signatures and key exchange. Interestingly, in spite of its prevalence, its implementation in OpenSSL has remained somewhat unoptimised, while less frequently used curves (P224, P256, P521) each have their own optimisations.
+
+The strategy I have used one might call a 56-bit redundant limb implementation of arithmetic modulo P384 with _Solinas reduction_. Without too much micro-optimisation, we get ~5.5x speedup over the default (Montgomery Multiplication) implementation on creating digital signatures.
```
Comment (ruscur) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: > we get ~5.5x speedup over the default (Montgomery Multiplication) implementation on creating digital signatures.

does "for" or "when" work better than "on" here?

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+How is this possible? Well first let's quickly explain some language:
+
+## Elliptic Curves
+
+When it comes to cryptography, it's more than likely someone with a computer science background will at least be familiar with ideas such as key-exchange and private-key signing. The stand-in asymetric cipher in a typical computer science curriculum will almost certainly by RSA. However the hayday of Elliptic Curve ciphers has well and truly arrived, and their operation seems no less mystical than when they were just a toy for academia.
```
Comment (ruscur) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: asymetric -> asymmetric

Comment (ruscur) [reply content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: hayday -> heyday

should there be a comma after "However"?

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+For example $\mathbb{F}_p$, that is the [unique](https://en.wikipedia.org/wiki/Finite_field#Existence_and_uniqueness) field with $p$ elements. The most straight-forward construction of this field is just to imagine arithmetic modulo $p$. The other finite fields used in practise in ECC are what some might call 'binary fields'. They're of the form $\mathbb{F}_{2^m}$, and will be familiar to you if you've studied AES (byte substitution is implemented by inversion modulo $\mathbb{F}_{2^8}$).
```
Comment (ruscur) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: >For example $\mathbb{F}_p$, that is

Could this just be 

For example $\mathbb{F}_p$ is

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+If we allow ourselves to be reductive for just a moment, you might observe that all ECC math for a particular curve therefore can be boiled down to just efficiently implementing fixed-point arithmetic modulo a single prime constant, $p$. From here on out, I'll be speaking from this abstraction layer alone. 101 level discrete maths should be the only prereq (hopefully).
+
+## Limbs
+
+You wish to multiply two $m$ bit numbers, have subdivided such a number into $n$ machine words in some way. Let's suppose just for now that $n$ divides $m$ neatly, then the quotient $d$ is the minimum number of bits in each machine word that will be required for representing our number. Suppose we use the straight-forward representation whereby the least significant $d$ bits are used for storing parts of our number, which we better call $x$ because this is crypto and descriptive variable names are deemed harmful apparently.
```
Comment (ruscur) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: >numbers, have

should this be "numbers and have"?  I wouldn't expect the comma unless there was another comma later in the sentence

>are deemed harmful apparently

I would use "are apparently deemed harmful" or "are deemed harmful (apparently)"

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+## Concerning Timing Side-Channels
+
+Cryptographic routines must perform all of their calculations in constant time. More specifically, it is important that timing cryptography code should not reveal any private keys or random nonces used during computation. Ultimately, all of our work so far has been to speed-up field arithmetic in the modulo field with prime $p_{384}$. But this is done in order to facilitate calculations in the secp384r1 elliptic curve, and ECDSA/ECDH each depend on being able to perform scalar 'point multiplication' (repeat application of the group operator). Since such an operation is inherently iterative, it presents the greatest potential for timing attacks.
+
+Simply put, we implement constant-time multiplication with the *wNAF* ladder method. This relies on pre-computing a window of multiples of the group generator, and then scaling and selectively adding multiples when required. [Wikipedia](https://en.wikipedia.org/wiki/Elliptic_curve_point_multiplication#Point_multiplication) provides a helpful primer to this method that cumulatively builds upon more naive approaches.
```
Comment (ruscur) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: I feel like there should be a conclusion of some sort.

**Review by ruscur — COMMENTED** [review]
Reviewed some english since I definitely can't review math

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown (hunk 4 above)
Comment (rohanmclure) [reply content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: Trying to quote Edger Dijkstra so I'll use the latter :P

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 6]
```diff
@@ -0,0 +1,218 @@
+Title: Going out on a Limb: Efficient Elliptic Curve Arithmetic in OpenSSL
+Authors: Rohan McLure
+Date: 2023-08-03 13:15
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: 04

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+From a performance perspective, great optimisations can be made by implementing efficient fixed-point arithmetic specialised to modulo by single prime constant, $p$. From here on out, I'll be speaking from this abstraction layer alone.
+
+## Limbs
+
+We wish to multiply two $m$ bit numbers and have subdivided such a number into $n$ machine words in some way. Let's suppose just for now that $n$ divides $m$ neatly, then the quotient $d$ is the minimum number of bits in each machine word that will be required for representing our number. Suppose we use the straight-forward representation whereby the least significant $d$ bits are used for storing parts of our number, which we better call $x$ because this is crypto and descriptive variable names are considered harmful (apparently).
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: "and have subdivided such a number" reads a bit weirdly, you're talking about two numbers then talking about one number or m-bit numbers in general or what?

Comment (ajdlinux) [reply content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: "machine words" might be slightly confusing terminology for an unfamiliar audience - give a concrete example ("64 bits on most architectures" or something)

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+$$x = \sum_{k = 0}^{n-1} 2^{dk} l_k$$
+
+If we then drop the requirement for each of our $n$ machine words (also referred to as a 'limb' from hereon out) to have no more than the least significant $d$ bits populated, we say that such an implementation uses 'redundant limbs', meaning that the $k$-th limb has high bits which overlap with the place values represented in the $(k+1)$-th limb.
+
+## Multiplication (mod p)
+
+The fundamental difficulty with making modulo arithmetic fast (and maintainingly cryptographic opacity) is to do with an annoying property of multiplication.
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: "maintaining".

what's "cryptographic opacity"?

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+But we can't keep doing our sums in redundant limb arithmetic forever, we must eventually reduce. Doing so may be expensive, and so we would rather reduce only when strictly necessary!
+
+## Solinas-ish Reduction
+
+Our prime is a _Solinas_ (_Pseudo/Generalised-Mersenne_) _Prime_. Mersenne Primes are primes expressible as $2^m - 1$. This can be generalised to low-degree polynomials in $2^m$. For example, another NIST curve uses $p_{224} = 2^{224} - 2^{96} + 1$ (a 224-bit number) where $p_{224} = f(2^{32})$ for $f(t) = t^7 - t^3 + 1$. The simpler the choice of polynomial, the simpler the modular reduction logic.
+
+Our choice of $t$ is $2^{56}$. Wikipedia provides a ideal example of the Solinas reduction algorithm for when the bitwidth of the prime is divisible by $\log_2{t}$, but that is not our scenario. We choose 56-bits for some pretty simple realities of hardware. 56 is less than 64 (standard machine word size) but not by too much, and the difference is byte-addressible ($64-56=8$). Let me explain:
+
+## Just the right amount of Reduction (mod p)
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: is there a reason why Reduction is capitalised?

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+Finally, let's reduce all the high bits of `in[6]`. Since `in[6]` has place value `t^6 = 2^{336}`, thus we wish to reduce all but the least significant $384 - 336 = 48$ bits.
+
+A goal in designing this algorithm is to ensure that `acc[6]` has as tight a bound as reasonably possible. Intuitively, if we can cause `acc[6]` to be as large as possible by absorbing the high bits of lower limbs, we reduce the number of bits that must be carried forward later on. As such, we perform a carry of the high-bits of `acc[4]`, `acc[5]` into `acc[6]` before we begin our substitution.
+
+Again, let
+
+$$X_3 = \sum^{5}_{k=0}\texttt{acc[k]}t^k + (\texttt{acc[6]} \text{(low bits)})t^6$$
+$$Y_3 = 2^{48}(\texttt{acc[6]} \text{(high bits, right shifted)}) t^6$$
+
+The equation for eliminating $2^{48}t^6$ is pretty straight forward:
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: straightforward

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            &+ (2^{124} - 2^{116} - 2^{68}) t \\
+            &+ (2^{124} + 2^{108} - 2^{76})
+\end{align*}
+$$
+
+We can then subtract values whose limbs are no larger than the least of these limbs above without fear of underflows providing us with an incorrect result. In our case, that upper bound for limb value is $2^{124} - 2^{116} - 2^{68} > 2^{123}$. Very comfortable.
+
+## Concerning Timing Side-Channels
+
+Cryptographic routines must perform all of their calculations in constant time. More specifically, it is important that timing cryptography code should not reveal any private keys or random nonces used during computation. Ultimately, all of our work so far has been to speed-up field arithmetic in the modulo field with prime $p_{384}$. But this is done in order to facilitate calculations in the secp384r1 elliptic curve, and ECDSA/ECDH each depend on being able to perform scalar 'point multiplication' (repeat application of the group operator). Since such an operation is inherently iterative, it presents the greatest potential for timing attacks.
```
Comment (ajdlinux) [inline content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: speed up

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown (hunk 9 above)
Comment (ajdlinux) [reply content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: should use consistent title or sentence case across all headings, pick either one as we don't have a strong tradition on this blogue

**Review by ajdlinux — COMMENTED** [review]
A few editorial comments, because I am too dumb to maths

File: content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown (hunk 9 above)
Comment (rohanmclure) [reply content/blog/2023-07-31-going-out-on-a-limb-efficient-ecc-p384.markdown]: Good to know. Switching to consistent title-case.

**Pull request merged**